Firestore/core/src/bundle/bundle_loader.cc

```diff
@@ -0,0 +1,135 @@
+#include "Firestore/core/src/model/document_key.h"
+#include "Firestore/core/src/model/document_key_set.h"
+#include "Firestore/core/src/model/model_fwd.h"
+#include "Firestore/core/src/model/no_document.h"
+
+namespace firebase {
+namespace firestore {
+namespace bundle {
+
+using firestore::Error;
+using firestore::api::LoadBundleTaskProgress;
+using firestore::api::LoadBundleTaskState;
+using model::DocumentKeySet;
+using model::MaybeDocumentMap;
+using model::NoDocument;
+using util::Status;
+using util::StatusOr;
+
+StatusOr<absl::optional<LoadBundleTaskProgress>> BundleLoader::AddElement(
+    std::unique_ptr<BundleElement> element_ptr, uint64_t byte_size) {
+  HARD_ASSERT(element_ptr->element_type() != BundleElement::Type::Metadata,
+              "Unexpected bundle metadata element.");
+
+  auto before_count = documents_.size();
+
+  const auto& element = *element_ptr;
+  if (element.element_type() == BundleElement::Type::NamedQuery) {
+    queries_.push_back(static_cast<const NamedQuery&>(element));
+  } else if (element.element_type() == BundleElement::Type::DocumentMetadata) {
+    const auto& document_metadata =
+        static_cast<const BundledDocumentMetadata&>(element);
+    current_document_ = document_metadata.key();
+    documents_metadata_.insert({document_metadata.key(), document_metadata});
+
+    if (!document_metadata.exists()) {
+      documents_ = documents_.insert(
+          document_metadata.key(),
+          NoDocument(document_metadata.key(), document_metadata.read_time(),
+                     /*has_committed_mutations=*/false));
+      current_document_ = absl::nullopt;
+    }
+  } else if (element.element_type() == BundleElement::Type::Document) {
+    const auto& document = static_cast<const BundleDocument&>(element);
+    if (!current_document_.has_value() ||
+        document.key() != current_document_.value()) {
+      return {util::Status::FromErrno(
+          Error::kErrorInvalidArgument,
+          "The document being added does not match the stored metadata.")};
+    }
+
+    documents_ = documents_.insert(document.key(), document.document());
+    current_document_ = absl::nullopt;
+  }
+
+  bytes_loaded_ += byte_size;
+
+  if (before_count == documents_.size()) {
+    return {absl::nullopt};
+  }
+
+  return {absl::make_optional(LoadBundleTaskProgress(
+      documents_.size(), metadata_.total_documents(), bytes_loaded_,
+      metadata_.total_bytes(), LoadBundleTaskState::kInProgress))};
+}
+
+StatusOr<MaybeDocumentMap> BundleLoader::ApplyChanges() {
+  if (current_document_ != absl::nullopt) {
+    return StatusOr<MaybeDocumentMap>(
+        Status::FromErrno(Error::kErrorInvalidArgument,
+                          "Bundled documents end with a document metadata "
+                          "element instead of a document."));
+  }
+  if (metadata_.total_documents() != documents_.size()) {
+    return StatusOr<MaybeDocumentMap>(Status::FromErrno(
```

Comment:
`FromErrno` is not applicable here, just use the `Status` constructor (applies throughout).

Reply:
Done.